### PR TITLE
New syntax for reversed structs

### PR DESCRIPTION
**Pull request by emma-miler**
This new file introduces a new syntax for defining structs we have reversed.
This allows us to define fields by their offset, and removes the need for padding fields, making them a lot more maintainable.

### CODE REVIEW

File: NorthstarDLL/filesystem.h [hunk 1]
```diff
@@ -5,25 +5,28 @@
...
+// clang-format off
+OFFSET_STRUCT(VPKFileEntry)
 {
-	char* directory;
-	char* filename;
-	char* extension;
-	unsigned char unknown[0x38];
+	STRUCT_SIZE(0x44);
+	FIELD(0x0, char* directory)
+	FIELD(0x4, char* filename)
```
Comment (BobTheBob9) [inline NorthstarDLL/filesystem.h:14]: 4 bytes for a ptr?

Comment (BobTheBob9) [reply NorthstarDLL/filesystem.h:14]: honestly i don't see much of a reason to explicitly define field offsets here, maybe a macro to define multiple fields inline from an offset (0 in this case) could be useful in situations where you don't need to do this? (assuming it isn't possible with the preexisting one

File: NorthstarDLL/structs.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+// To use these macros, define a struct like so:
+/*
+OFFSET_STRUCT(Name)
+{
+	STRUCT_SIZE(0x100)		// Total in-memory struct size
+	FIELD(0x0, int field)	// offset, signature
+}
+*/
+
+#define OFFSET_STRUCT(name) union name
```
Comment (BobTheBob9) [inline NorthstarDLL/structs.h:26]: debating whether this should be called something like INTERNAL_STRUCT instead, but unsure

**Review by BobTheBob9 — CHANGES_REQUESTED** [review]
looks great mostly

File: NorthstarDLL/filesystem.h (hunk 1 above)
Comment (emma-miler) [reply NorthstarDLL/filesystem.h:14]: This is fair actually. I could revert this to its old def

File: NorthstarDLL/structs.h (hunk 2 above)
Comment (emma-miler) [reply NorthstarDLL/structs.h:26]: Well its not really internal or anything. OFFSET_STRUCT describes what it is: a struct defined with offsets

Comment (BobTheBob9) [reply NorthstarDLL/structs.h:26]: yeah fair

**Review by BobTheBob9 — APPROVED** [review]

**Pull request merged**